libkbfs/md_util.go

```diff
@@ -377,6 +379,116 @@ func signMD(
 	return nil
 }
 
+func getFileBlockForMD(ctx context.Context, config Config, ptr BlockPointer,
+	rmdToDecrypt *RootMetadata, rmdWithKeys KeyMetadata) (*FileBlock, error) {
+	// We don't have a convenient way to fetch the block from here via
+	// folderBlockOps, so just go directly via the
+	// BlockCache/BlockOps.  No locking around the blocks is needed
+	// since these change blocks are read-only.
+	block, err := config.BlockCache().Get(ptr)
+	if err != nil {
+		block = NewFileBlock()
+		if err := config.BlockOps().Get(
+			ctx, rmdWithKeys, ptr, block); err != nil {
+			return nil, err
+		}
+		if err := config.BlockCache().Put(
+			ptr, rmdToDecrypt.TlfID(), block, TransientEntry); err != nil {
+			return nil, err
+		}
+	}
+
+	fblock, ok := block.(*FileBlock)
+	if !ok {
+		return nil, NotFileBlockError{ptr, MasterBranch, path{}}
+	}
+	return fblock, nil
+}
+
+func reembedBlockChanges(ctx context.Context, config Config,
+	rmdToDecrypt *RootMetadata, rmdWithKeys KeyMetadata) error {
+	info := rmdToDecrypt.data.Changes.Info
+	if info.BlockPointer == zeroPtr {
+		return nil
+	}
+
+	// Fetch the top-level block.
+	fblock, err := getFileBlockForMD(
+		ctx, config, info.BlockPointer, rmdToDecrypt, rmdWithKeys)
+	if err != nil {
+		return err
+	}
+
+	buf := fblock.Contents
+	if fblock.IsInd {
+		numFetchers := len(fblock.IPtrs)
+		if numFetchers > maxParallelBlockGets {
+			numFetchers = maxParallelBlockGets
+		}
+
+		// We don't yet know the plaintext size of the buffer, so size
+		// it to be the size of the first n-1 child blocks, and we'll
+		// append the final block to it once we have it.
+		lastOff := fblock.IPtrs[len(fblock.IPtrs)-1].Off
+		var bufLock sync.Mutex
+		buf = make([]byte, lastOff, lastOff*2)
+
+		// Fetch all the child blocks in parallel.
+		iptrsToFetch := make(chan IndirectFilePtr, len(fblock.IPtrs))
+		eg, groupCtx := errgroup.WithContext(ctx)
+		fetchFn := func() error {
+			for iptr := range iptrsToFetch {
+				select {
+				case <-groupCtx.Done():
+					return groupCtx.Err()
+				default:
+				}
+
+				fblock, err := getFileBlockForMD(groupCtx, config,
+					iptr.BlockPointer, rmdToDecrypt, rmdWithKeys)
+				if err != nil {
+					return err
+				}
+
+				func() {
```

Comment:
You can even size buf exactly by draining `indirectBlocks` into a slice first and looking for the last block. Guessing at the capacity makes me slightly nervous of overflow, although it may not matter if we can assume 64-bit ints.